addons/web_editor/static/lib/odoo-editor/src/commands/commands.js

```diff
@@ -48,75 +48,130 @@ const BG_CLASSES_REGEX = /\bbg-[^\s]*\b/g;
 function insert(editor, data, isText = true) {
     const selection = editor.document.getSelection();
     const range = selection.getRangeAt(0);
-    let startNode;
+    let currentNode;
     let insertBefore = false;
     if (selection.isCollapsed) {
         if (range.startContainer.nodeType === Node.TEXT_NODE) {
             insertBefore = !range.startOffset;
             splitTextNode(range.startContainer, range.startOffset, DIRECTIONS.LEFT);
-            startNode = range.startContainer;
+            currentNode = range.startContainer;
         }
     } else {
         editor.deleteRange(selection);
     }
-    startNode = startNode || editor.document.getSelection().anchorNode;
-    if (startNode.nodeType === Node.ELEMENT_NODE) {
+    currentNode = currentNode || editor.document.getSelection().anchorNode;
+    if (currentNode.nodeType === Node.ELEMENT_NODE) {
         if (selection.anchorOffset === 0) {
-            startNode.prepend(editor.document.createTextNode(''));
-            startNode = startNode.firstChild;
+            currentNode.prepend(editor.document.createTextNode(''));
+            currentNode = currentNode.firstChild;
         } else {
-            startNode = startNode.childNodes[selection.anchorOffset - 1];
+            currentNode = currentNode.childNodes[selection.anchorOffset - 1];
         }
     }
 
     const fakeEl = document.createElement('fake-element');
+    const fakeElFirstChild = document.createElement('fake-element-fc');
+    const fakeElLastChild = document.createElement('fake-element-lc');
     if (isText) {
         fakeEl.innerText = data;
     } else {
         fakeEl.innerHTML = data;
     }
+
+    // In case the html inserted is all contained in a single root <P> tag,
+    // we take the all content of the <p> and avoid inserting the <p>.
+    if (fakeEl.firstChild === fakeEl.lastChild && fakeEl.firstChild.nodeName === 'P') {
+        const p = fakeEl.firstElementChild;
+        fakeEl.innerHTML = p.innerHTML;
```

Comment:
This kind of stuff is triggering the security ci. It is not clear from the changes in this PR where the sanitization is done if there is any. I actually suspect that there isn't. Maybe we could use DOMPurify like we do in collaborative mode ?

Reply:
I think we can achieve this efficiently enough without `innerHTML`, like:
```js
const p = fakeEl.firstChild.cloneNode(true);
[...fakeEl.childNodes].forEach(child => child.remove());
fakeEl.append(...p.childNodes);
```

Reply:
You mean `replaceChildren` ? 😎

Reply:
Holy forking shirtballs that [exists](https://developer.mozilla.org/en-US/docs/Web/API/Element/replaceChildren)! Cool! Yes so then
```js
fakeEl.replaceChildren(...fakeEl.firstChild.cloneNode(true).childNodes);
```
Maybe we don't even need the `cloneNode(true)` if `replaceChildren` is smart enough.

Side node, "replace children" sounds like a command to a dystopian future's household robot for when your kids spilled a paint bucket all over your new kitchen's floor.

Reply:
> This kind of stuff is triggering the security ci. It is not clear from the changes in this PR where the sanitization is done if there is any. I actually suspect that there isn't. Maybe we could use DOMPurify like we do in collaborative mode ?

The HTML data arriving here is in fact "sanitized" in the `_cleanForPaste()` method of the OdooEditor.

https://github.com/odoo-dev/odoo/blob/15.0-web-editor-paste-fix-4-sge/addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js#L2106

Reply:
Just for the record, I compared the speed of `replaceChildren` with that of `innerHTML =` and `replaceChildren` is a lot faster. https://jsfiddle.net/Zinston/9j6tmcua/7/ 

Sound like the household robot made an efficient choice.

Reply:
Will adapt